File: backend/src/inference/eval/toolcall/matrix.rs
use crate::errors::AppResult;
use crate::inference::backend::backend_kind::BackendKind;
use crate::inference::eval::toolcall::eval::ToolCallReport;
use crate::persistence::eval_history::RunSummary;
use serde::{Deserialize, Serialize};

/// One model+backend to run a collection against (sent by the frontend; also
/// persisted in the resumable job-queue header, hence `Serialize`).
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ModelTarget {
    pub model: String,
    pub backend: BackendKind,
    /// Reasoning model (the sidebar "thinking" checkbox). Drives the raised per-turn token
    /// budget + `<think>` stripping in the agentic runner. `#[serde(default)]` so a job log
    /// written before this field (or a non-thinking target) deserializes as `false`.
    #[serde(default)]
    pub is_thinking: bool,
}

/// One model's result for the whole collection: either a full report or the
/// error that model hit (a down backend must not fail the rest of the matrix).
#[derive(Serialize, Clone, Debug, PartialEq)]
pub struct MatrixColumn {
    pub model: String,
    pub backend: BackendKind,
    pub report: Option<ToolCallReport>,
    pub error: Option<String>,
}

/// The full matrix: one column per target plus the mean composite across the
/// columns that succeeded.
#[derive(Serialize, Clone, Debug, PartialEq)]
pub struct MatrixReport {
    pub collection_id: String,
    pub columns: Vec<MatrixColumn>,
    pub avg_score: Option<f64>,
}

/// Fold per-target run outcomes into a matrix report. Pure (no async / no I/O) so
/// the mapping + averaging is unit-testable without a model server.
pub fn build_matrix(
    collection_id: &str,
    results: Vec<(ModelTarget, AppResult<ToolCallReport>)>,
) -> MatrixReport {
    let columns: Vec<MatrixColumn> = results
        .into_iter()
        .map(|(target, res)| match res {
            Ok(report) => MatrixColumn { model: target.model, backend: target.backend, report: Some(report), error: None },
            Err(e) => MatrixColumn { model: target.model, backend: target.backend, report: None, error: Some(e.to_string()) },
        })
        .collect();

    let scored: Vec<f64> = columns
        .iter()
        .filter_map(|c| c.report.as_ref().and_then(|r| r.composite))
        .collect();
    let avg_score = if scored.is_empty() { None } else { Some(scored.iter().sum::<f64>() / scored.len() as f64) };

    MatrixReport { collection_id: collection_id.to_string(), columns, avg_score }
}

/// One history `RunSummary` per successful column, stamped with `ts`. Failed
/// columns are not recorded (only real measurements enter the regression log).
pub fn summaries(report: &MatrixReport, ts: &str) -> Vec<RunSummary> {
    report
        .columns
        .iter()
        .filter_map(|c| {
            let r = c.report.as_ref()?;
            Some(RunSummary {
                ts: ts.to_string(),
                model: c.model.clone(),
                backend: c.backend,
                parse_rate: r.parse_rate,
                tool_selection_acc: r.tool_selection_acc,
                arg_acc: r.arg_acc,
                abstain_acc: r.abstain_acc,
                composite: r.composite,
                n: r.n,
                pass_k: None,
                agentic_avg_steps: None,
                effort: None,
                is_thinking: false, // single-turn matrix: no agentic effort, thinking N/A
            })
        })
        .collect()
}

#[cfg(test)]
#[path = "matrix_tests.rs"]
mod tests;

File: frontend/src/shared/ipc/eval/matrix.ts
import { invoke } from "@tauri-apps/api/core";
import { z } from "zod";
import { BackendKindSchema, type BackendKind } from "../models/storage";
import { ToolCallReportSchema } from "./toolcall";
import type { ToolTask } from "./registry";

/// One model+backend to run a collection against.
export interface ModelTarget {
  model: string;
  backend: BackendKind;
  /// Reasoning model (sidebar "thinking" toggle): the backend raises its per-turn token
  /// budget and strips <think> before scoring. Optional — omitted/false = a terse model.
  is_thinking?: boolean;
}

/// One model's outcome for the whole collection — a report, or the error it hit.
export const MatrixColumnSchema = z.object({
  model: z.string(),
  backend: BackendKindSchema,
  report: ToolCallReportSchema.nullable(),
  error: z.string().nullable(),
});
export type MatrixColumn = z.infer<typeof MatrixColumnSchema>;

export const MatrixReportSchema = z.object({
  collection_id: z.string(),
  columns: z.array(MatrixColumnSchema),
  avg_score: z.number().nullable(),
});
export type MatrixReport = z.infer<typeof MatrixReportSchema>;

/// One recorded run for the regression timeline.
export const RunSummarySchema = z.object({
  ts: z.string(),
  model: z.string(),
  backend: BackendKindSchema,
  parse_rate: z.number().nullable(),
  tool_selection_acc: z.number().nullable(),
  arg_acc: z.number().nullable(),
  abstain_acc: z.number().nullable(),
  composite: z.number().nullable(),
  n: z.number().int().nonnegative(),
  // Agentic metrics (Phase 6) — absent in pre-Phase-6 history, so optional.
  pass_k: z.number().nullable().optional(),
  agentic_avg_steps: z.number().nullable().optional(),
  effort: z.number().nullable().optional(),
});
export type RunSummary = z.infer<typeof RunSummarySchema>;

/// Batch-run a collection's tasks against several models (sequential, backend
/// side) and get a per-model matrix; successful columns are logged to history.
export async function runCollectionMatrix(
  collectionId: string,
  targets: ModelTarget[],
  tasks: ToolTask[],
): Promise<MatrixReport> {
  return MatrixReportSchema.parse(
    await invoke("run_collection_matrix", { collectionId, targets, tasks }),
  );
}

/// The recorded run history for a collection, oldest first.
export async function loadCollectionHistory(collectionId: string): Promise<RunSummary[]> {
  return z.array(RunSummarySchema).parse(await invoke("load_collection_history", { collectionId }));
}
